lib/Doctrine/DBAL/ForwardCompatibility/Result.php

```diff
@@ -301,4 +303,104 @@ private function ensureHasKeyValue(): void
             throw NoKeyValue::fromColumnCount($columnCount);
         }
     }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @deprecated This feature will no longer be available on Result object in 3.0.x version.
+     */
+    public function bindValue($param, $value, $type = ParameterType::STRING)
+    {
+        Deprecation::triggerIfCalledFromOutside(
+            'doctrine/dbal',
+            'https://github.com/doctrine/dbal/pull/4019',
+            'Result::bindValue() is deprecated, no replacement.'
+        );
+
+        if ($this->stmt instanceof DriverStatement) {
+            return $this->stmt->bindValue($param, $value, $type);
+        }
+
+        throw Exception::notSupported('bindValue');
```

Comment:
I'm getting a bit lost: the forward-compatible result implements the statement interface and can act as a statement depending on the wrapped implementation. @beberlei could you review this please?

Reply:
In 2.12.x version :
- Doctrine\DBAL\Connection::executeCacheQuery() returned a Driver\ResultStatement object
- Doctrine\DBAL\Connection::executeQuery() returned a Driver\Statement object, except when using cache where a Driver\ResultStatement is returned instead

The solution makes both return a result implementing Driver\Statement, but Statement feature will only be available when using Doctrine\DBAL\Connection::executeQuery().

We can be more specific in methods @return documentation by adding following interfaces :
- ForwardCompatibility\DriverResultStatement implementing Driver\ResultStatement&DBAL\Result for executeCacheQuery()
- ForwardCompatibility\DriverStatement implementing Driver\Statement&DBAL\Result for executeQuery()

Reply:
Thank you for the explanation. Makes sense. I think it's enough to leave the return type documented via an annotation with the intersection type, no need to introduce more interfaces.

Reply:
I can't find a way to fix documentation in QueryBuilder without adding an interface : https://github.com/doctrine/dbal/pull/4596/files#diff-829b37fd7ed65c7ead12002f3e689f3a672986dd88fd0f5d4d5f92321a61211dL204

Without interface, we will have to write something like :
```php
    /**
     * Executes this query using the bound parameters and their types.
     *
     * @return (ResultStatement&Doctrine\DBAL\Result)|int
     *
     * @throws Exception
     */
```

I test it with phpstan but it does not work.

Reply:
Also, I think we should to be more accurate on Doctrine\DBAL\Connection::executeQuery() with something like :

```php
    /**
     * ...
     * @return (Driver\ResultStatement&Doctrine\DBAL\Result)|(Driver\Statement&Doctrine\DBAL\Result) Return the executed result statement or cached statement if a cache profile is used
     *
     * @throws Exception
     */
```
Same issue here, I don't think phpstan can interpret complex type like that.

Reply:
I see. Then the interfaces are necessary. The question is, is there a way to test the newly introduced types? Currently, the build passes but there are changes needed to satisfy the static analysis on the consumer side. That means that we're not analyzing all the cases ourselves. E.g. it might help to enable phpstan for some of the test cases that describe the consumers of the old API and make sure that such analysis passes.

Reply:
That's the meaning of tests added here : https://github.com/doctrine/dbal/pull/4596/files#diff-1056c8f8b7e1dbbb024efeed5bca59a3e9097603fafb63d0c7c7924531b2d63eR363-R381

But I still have to find a way to test executeQuery() with cache param.

Reply:
I think this is correct, it exposes the `bind*` api from the result, but since it is still acting as a statement in DBAL 2 this is correct and the deprecations trigger correctly. I believe we can roll with this.